wpimath/src/main/java/edu/wpi/first/math/estimator/SwerveDrivePoseEstimator.java

```diff
@@ -177,42 +205,43 @@ public Pose2d getEstimatedPosition() {
    *     or sync the epochs.
    */
   public void addVisionMeasurement(Pose2d visionRobotPoseMeters, double timestampSeconds) {
-    // Step 1: Get the pose odometry measured at the moment the vision measurement was made.
+    // Step 1: Get the odometry pose measured at the moment the vision measurement was made.
     var sample = m_poseBuffer.getSample(timestampSeconds);
 
     if (sample.isEmpty()) {
       return;
     }
 
-    // Step 2: Measure the twist between the odometry pose and the vision pose.
-    var twist = sample.get().poseMeters.log(visionRobotPoseMeters);
+    // Step 2: Get the odometry delta in the time since the vision measurement,
+    var delta = new Transform2d(sample.get().poseMeters, m_odometry.getPoseMeters());
+    // and apply its inverse to the current pose estimate to get the estimate at the time
+    // of the vision measurement
+    var old_estimate = m_poseEstimate.transformBy(delta.inverse());
+
+    // Step 3: Measure the twist between the "old" estimated pose and the vision pose
+    var twist = old_estimate.log(visionRobotPoseMeters);
 
-    // Step 3: We should not trust the twist entirely, so instead we scale this twist by a Kalman
+    // Step 4: We should not trust the twist entirely, so instead we scale this twist by a Kalman
     // gain matrix representing how much we trust vision measurements compared to our current pose.
     var k_times_twist = m_visionK.times(VecBuilder.fill(twist.dx, twist.dy, twist.dtheta));
 
-    // Step 4: Convert back to Twist2d.
-    var scaledTwist =
-        new Twist2d(k_times_twist.get(0, 0), k_times_twist.get(1, 0), k_times_twist.get(2, 0));
+    // Step 5: The Kalman gains scale the vision twists independent of other vision measurements,
+    // so applying multiple vision measurements before the next update will have different results
+    // depending on the order they are applied in. We can sum the Kalman gains applied before the
+    // next update and use that to effectively average the twist for multiple vision measurements.
+    var weighted_k_times_twist =
+        k_times_twist.elementTimes(m_visionKSum.diag().extractColumnVector(0).elementPower(-1));
```

Comment:
Is this the mathematically optimal estimation result though? Averaging the Kalman gain is very suspicious.

Reply:
I have no clue what is considered mathematically optimal here.

The motivation for this is that teams may be adding multiple vision pose estimates at once when multiple AprilTags are visible, but the order these are applied in matters as described above. Later measurements are effectively weighted more, so applying two vision measurements `(0,1,0)` and `(0,-1,0)` with equal standard deviations when the current pose estimate is `(0,0,0)` will cause a non-zero correction towards the last vision measurement applied.

Running the described example with default standard deviations:
```
Current wpilib pose estimate: (0.0, -0.05263157894736839, 0.0)
This PR: (0.0, -9.215718466126788E-18, 0.0)
```

The use of the Kalman gain sum matrix in this PR is the same as (weighted) averaging all the vision measurements applied before the next odometry update. From my admittedly limited understanding of the Kalman filter, this is similar to how the Kalman gain K would change over time as the error covariance P changes when new measurements are added (which does not happen in the current implementation).

If you'd like, I can separate this functionality into a separate PR.

Reply:
> Later measurements are effectively weighted more, so applying two vision measurements (0,1,0) and (0,-1,0) with equal standard deviations when the current pose estimate is (0,0,0) will cause a non-zero correction towards the last vision measurement applied.

Later measurements being weighted more is completely intentional. That's how Kalman filters work.

> From my admittedly limited understanding of the Kalman filter, this is similar to how the Kalman gain K would change over time as the error covariance P changes when new measurements are added (which does not happen in the current implementation).

False. For linear models (which we are using), the error covariance and Kalman gain converge to a steady-state value.

> If you'd like, I can separate this functionality into a separate PR.

The CME fix should be in a separate PR. This PR does several things and has a really vague title as a result.